### PR TITLE
Code improvements.

### PR DESCRIPTION
**Pull request by Martinski4GitHub**
Code improvements in the "`_GetAllNodeSettings_` " function.

### CODE REVIEW

**Comment by Martinski4GitHub** [comment]
@ExtremeFiretop,

Whenever you can, please review & verify the PR changes. I decided to submit this PR separately before I started working on the new menu option to view the logs. On that topic, I'm planning to show the list of the most recent log files up to a maximum of 20 and let the user select which one to view. I believe the last 20 should be enough, even if users are running the cron job daily (or perhaps more frequently than that?). Any objections? Other suggestions?

**Comment by ExtremeFiretop** [comment]
> @ExtremeFiretop,
> 
> Whenever you can, please review & verify the PR changes. I decided to submit this PR separately before I started working on the new menu option to view the logs. On that topic, I'm planning to show the list of the most recent log files up to a maximum of 20 and let the user select which one to view. I believe the last 20 should be enough, even if users are running the cron job daily (or perhaps more frequently than that?). Any objections? Other suggestions?

Actually yes I was thinking about this today.
First of all, I like the idea of the last 20 because we save the last 3 months lol. So I think the last 20 is good. Even if someone runs the cron often it should good them a good view of the past without crowding the shell 

Also, I think I want to include a way to view the changelog file for the firmware as well.

And considering we will be adding this to view the log files and we have an option to move the log file location, how do you feel about creating a new menu all together to encapsulate all these log options? 

Basically like a "Log Menu" where you can set the log directory, view existing logs, and possibly the firmware changelogs if high risk phrases are found.

Any thoughts on that?

**Comment by ExtremeFiretop** [comment]
@Martinski4GitHub buddy I'm back!
I got a few minutes until dinner so I'm gonna test this now, doesn't look like it should take too long.

File: MerlinAU.sh [hunk 1]
```diff
@@ -3677,7 +3671,7 @@ Please manually update to version $minimum_supported_version or higher to use th
 
         if [ ! -f "$changeLogFile" ]
         then
-            Say "Change-log file [$changeLogFile] does NOT exist."
+            Say "Change-log file [${FW_BIN_DIR}/$Changelog-${changeLogTag}.txt] does NOT exist."
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:3674]: Approved. Smart call I actually forgot about this from his screenshot but it said blank file was missing lol!

File: MerlinAU.sh [hunk 2]
```diff
@@ -3107,7 +3101,7 @@ _CheckNodeFWUpdateNotification_()
            nodefwNewUpdateNotificationVers="$2"
            nodefwNewUpdateNotificationDate="$(date +"$FW_UpdateNotificationDateFormat")"
            _Populate_Node_Settings_ "$node_label_mac" "$node_lan_hostname" "$nodefwNewUpdateNotificationDate" "$nodefwNewUpdateNotificationVers" "$uid"
-           nodefriendlyname="$(_GetAllNodeSettings_ "$node_label_mac" "NameID")"
+           nodefriendlyname="$(_GetAllNodeSettings_ "$node_label_mac" "Model_NameID")"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:3104]: Approved and tested with success.

File: MerlinAU.sh [hunk 3]
```diff
@@ -790,37 +790,31 @@ Get_Custom_Setting()
...
 _GetAllNodeSettings_()
 {
-    if [ $# -lt 2 ] || [ -z "$1" ] || [ -z "$2" ]; then return 1; fi
+    if [ $# -lt 2 ] || [ -z "$1" ] || [ -z "$2" ]
+    then echo "**ERROR**" ; return 1; fi
 
-    local node_mac_address="$1"  pattern
-    local setting_value  setting_part="$2"  matched_lines
-    local default_value="TBD"
+    ## Node Setting KEY="Node_{MACaddress}_{keySuffix}" ##
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:801]: Approved and tested with success :)
Thank you for the additional cleanup! Your touch is always appreciated

File: MerlinAU.sh [hunk 4]
```diff
@@ -2281,21 +2275,21 @@ _NodeActiveStatus_()
...
     local update_vers="$4"
-    local node_suffix="$mac_address"
-    local node_prefix="Node_"
+    local nodeKeyPrefix="Node_${MAC_address}_"
 
     # Update or add each piece of information
-    Update_Custom_Settings "${node_prefix}${node_suffix}_Model_NameID" "$model_id"
-    Update_Custom_Settings "${node_prefix}${node_suffix}_New_Notification_Date" "$update_date"
-    Update_Custom_Settings "${node_prefix}${node_suffix}_New_Notification_Vers" "$update_vers"
+    Update_Custom_Settings "${nodeKeyPrefix}Model_NameID" "$model_id"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:2290]: Also approved.

**Review by ExtremeFiretop — APPROVED** [review]

**Comment by Martinski4GitHub** [comment]
> 
> Actually yes I was thinking about this today. First of all, I like the idea of the last 20 because we save the last 3 months lol. So I think the last 20 is good. Even if someone runs the cron often it should good them a good view of the past without crowding the shell
> 
3 months!! AFAIK, there's code in the script that deletes log files older than 30 days. Are you going to change that?

>
> Also, I think I want to include a way to view the changelog file for the firmware as well.
> 
I'm not sure I fully understand. Do you want to add a menu option to toggle viewing the Changelog if "high-risk phrases" are found **after** the F/W zip file is downloaded but **before** the FW/ image is flashed?

Am I getting this correctly? If not, can you elaborate on what the workflow would be?

>
> And considering we will be adding this to view the log files and we have an option to move the log file location, how do you feel about creating a new menu all together to encapsulate all these log options?
> 
> Basically like a "Log Menu" where you can set the log directory, view existing logs, and possibly the firmware changelogs if high risk phrases are found.
> 
First, I'd like to understand your train of thought about viewing the changelogs before commenting on this. I don't have quite the "full picture" yet to make an informed response.

**Comment by ExtremeFiretop** [comment]
> > Actually yes I was thinking about this today. First of all, I like the idea of the last 20 because we save the last 3 months lol. So I think the last 20 is good. Even if someone runs the cron often it should good them a good view of the past without crowding the shell
> 
> 3 months!! AFAIK, there's code in the script that deletes log files older than 30 days. Are you going to change that?
> 
> > Also, I think I want to include a way to view the changelog file for the firmware as well.
> 
> I'm not sure I fully understand. Do you want to add a menu option to toggle viewing the Changelog if "high-risk phrases" are found **after** the F/W zip file is downloaded but **before** the FW/ image is flashed?
> 
> Am I getting this correctly? If not, can you elaborate on what the workflow would be?
> 
> > And considering we will be adding this to view the log files and we have an option to move the log file location, how do you feel about creating a new menu all together to encapsulate all these log options?
> > Basically like a "Log Menu" where you can set the log directory, view existing logs, and possibly the firmware changelogs if high risk phrases are found.
> 
> First, I'd like to understand your train of thought about viewing the changelogs before commenting on this. I don't have quite the "full picture" yet to make an informed response.

If the code says 30 days you must be right, I was just going from memory when I wrote that.

As for your question I think the workflow would look something like:

If high risk phrases are found.
And a user selects to cancel the flash.
Save the changelog file from deletion.
If the changelog file exists. Display an option to view it under this new "Log Menu".

**Comment by ExtremeFiretop** [comment]
Another alternative:

If high risk phrases are found, send the content of the scanned log file in the email notification?

I'm just spit balling design ideas here.

**Comment by ExtremeFiretop** [comment]
FYI I'm just sitting down for dinner in front of the TV I'll be back in a few

**Comment by Martinski4GitHub** [comment]
> > 
> > 
> > First, I'd like to understand your train of thought about viewing the changelogs before commenting on this. I don't have quite the "full picture" yet to make an informed response.
> 
> If the code says 30 days you must be right, I was just going from memory when I wrote that.
> 
> As for your question I think the workflow would look something like:
> 
> If high risk phrases are found. 
> And a user selects to cancel the flash. 
> Save the changelog file from deletion. 
> If the changelog file exists. Display an option to view it under this new "Log Menu".
>
If I understand the desired end-result, wouldn't it be simpler to just provide a menu option to download the appropriate Changelog "on demand" **without** actually downloading any F/W ZIP file?

For  example, we add a new menu option like:
`
cl) View the latest F/W Changelog file.
`

When the user enters "**cl**" the script would download the appropriate data (via **curl** cmd) using the corresponding URL.

Changelog-386.txt:
https://www.asuswrt-merlin.net/node/10

Changelog-NG.txt:
https://www.asuswrt-merlin.net/changelog

Then the script extracts the "**Changelog-xxx.file.txt**" data and presents it to the user for review.
This menu option would be available all the time, and the user can select to view the latest log at any time, whether or not it contains "high-risk phrases."

**Comment by Martinski4GitHub** [comment]
> Another alternative:
> 
> If high risk phrases are found, send the content of the scanned log file in the email notification?
> 
In this case, the email notification could include the appropriate URL of the corresponding Changelog.

Changelog-386.txt:
https://www.asuswrt-merlin.net/node/10

Changelog-NG.txt:
https://www.asuswrt-merlin.net/changelog

Sometimes simpler is better.

**Comment by ExtremeFiretop** [comment]
> > Another alternative:
> > If high risk phrases are found, send the content of the scanned log file in the email notification?
> 
> In this case, the email notification could include the appropriate URL of the corresponding Changelog.
> 
> Changelog-386.txt: https://www.asuswrt-merlin.net/node/10
> 
> Changelog-NG.txt: https://www.asuswrt-merlin.net/changelog
> 
> Sometimes simpler is better.

Simplest thing I think would be doing what I mentioned, right?:

```
       STOP_FW_UPDATE_APPROVAL)
           emailBodyTitle="WARNING"
           {
             echo "Found high-risk phrases in the change-logs while Auto-Updating to version <b>${fwNewUpdateVersion}</b> on the <b>${MODEL_ID}</b> router."
             printf "\nPlease run script interactively to approve this F/W Update from current version:\n<b>${fwInstalledVersion}</b>\n"
           } > "$tempEMailBodyMsg"
           ;;
```

Take this code, and add this: $changelog_contents somewhere in it? Right?

Ends up looking like:

```
       STOP_FW_UPDATE_APPROVAL)
           emailBodyTitle="WARNING"
           {
             echo "Found high-risk phrases in the change-logs while Auto-Updating to version <b>${fwNewUpdateVersion}</b> on the <b>${MODEL_ID}</b> router."
             echo "Changelog contents include the following changes: $changelog_contents"
             printf "\nPlease run script interactively to approve this F/W Update from current version:\n<b>${fwInstalledVersion}</b>\n"
           } > "$tempEMailBodyMsg"
           ;;
```

Really that's all I was asking for :) No need to download anything on demand, no need to include any URL's in the emails :)
All that requires extra code, this above really only requires on additional value

**Comment by ExtremeFiretop** [comment]
With the above change, I could see why it failed and make my own choice/decision before even running MerlinAU.
I also wouldn't need to go above and beyond to manually download the changelog and review it manually. :)

**Comment by ExtremeFiretop** [comment]
> With the above change, I could see why it failed and make my own choice/decision before even running MerlinAU. I also wouldn't need to go above and beyond to manually download the changelog and review it manually. :)

Also it would of already sorted and only kept the propriate changes between the current and upcoming firmware, instead of all of the changes since the dawn of time haha

**Comment by ExtremeFiretop** [comment]
@Martinski4GitHub 

I'm having a coffee and thinking, there is a limitation to doing it that way though.
Even though it's the simplest, it does require you to have email notifications setup in AMTM and MerlinAU.

So naturally there would be that dependency else you'd still have to manually download to review the changelogs if high risk phrases are found.

**Comment by ExtremeFiretop** [comment]
I'm thinking I'll let you start how you see fit for the Logs Menu.
We can always modify and adjust later, I'll keep sleeping on the changelog idea.

**Comment by ExtremeFiretop** [comment]
Any ideas or thoughts on this one: https://www.snbforums.com/threads/merlinau-v1-1-2-the-ultimate-firmware-auto-updater-now-available-in-amtm.88577/post-904825

**Comment by ExtremeFiretop** [comment]
> > > First, I'd like to understand your train of thought about viewing the changelogs before commenting on this. I don't have quite the "full picture" yet to make an informed response.
> > 
> > 
> > If the code says 30 days you must be right, I was just going from memory when I wrote that.
> > As for your question I think the workflow would look something like:
> > If high risk phrases are found.
> > And a user selects to cancel the flash.
> > Save the changelog file from deletion.
> > If the changelog file exists. Display an option to view it under this new "Log Menu".
> 
> If I understand the desired end-result, wouldn't it be simpler to just provide a menu option to download the appropriate Changelog "on demand" **without** actually downloading any F/W ZIP file?
> 
> For example, we add a new menu option like: `cl) View the latest F/W Changelog file.`
> 
> When the user enters "**cl**" the script would download the appropriate data (via **curl** cmd) using the corresponding URL.
> 
> Changelog-386.txt: https://www.asuswrt-merlin.net/node/10
> 
> Changelog-NG.txt: https://www.asuswrt-merlin.net/changelog
> 
> Then the script extracts the "**Changelog-xxx.file.txt**" data and presents it to the user for review. This menu option would be available all the time, and the user can select to view the latest log at any time, whether or not it contains "high-risk phrases."

The more I think about it, the more I like this solution even though it isn't simpler, why? because it's available all the time, even if the user hasn't setup email notifications. That removes the dependency to force people to set that up to review the changelogs.

**Comment by Martinski4GitHub** [comment]
> 
> Simplest thing I think would be doing what I mentioned:
> 
> ...
> 
> Really that's all I was suggesting for changelogs :) No need to download anything on demand, no need to include any URL's in the emails :) All that requires extra code, this above really only requires one additional value
>
...
>
> With the above change, I could see why it failed and make my own choice/decision before even running MerlinAU. I also wouldn't need to go above and beyond to manually download the changelog and review it manually. :)
>
...
> 
> 
> Also it would of already sorted and only kept the propriate changes between the current and upcoming firmware, instead of all of the changes since the dawn of time haha
>


You seemed to have conflated the two separate features/events.

The new "`cl) View the latest F/W Changelog file.`" menu option implementation (as I suggested above) would be available all the time, regardless of any “high-risk phrases” and **independently** of any existing email notification setup that informs users of such “high-risk phrases” ***after*** trying to flash the FW image. The menu option would allow users to review the Changelog at any time on-demand, before or after deciding to flash F/W image, and whether email notifications are set up by the user.

The 2nd feature/event is the email notification itself, which happens only **after** trying to flash the F/W image & only if email notifications are set up. During this event, the above menu option has nothing to do with the email. They are 2 completely separate & independent features, related only because they deal with the Changelog. Therefore, the notification email doesn’t need to download (again) the Changelog on-demand, so it would be simpler to provide the Changelog URL in the email itself so users can just click on it to review the latest changes listed at the top of the page. Yes, it includes previous versions, but that’s not a “bad” thing as it gives you more context of what came before & can compare the currently installed version vs the next update version (before deciding to continue to flash the new FW image).

Do you see what I mean by the 2 features/events being separate & independent of each other? 
In fact, you can modify the email notification contents now as you see fit without having the new menu option implemented.

We can talk more later tonight but I hope things are much clearer now.

**Comment by ExtremeFiretop** [comment]
> > Simplest thing I think would be doing what I mentioned:
> > ...
> > Really that's all I was suggesting for changelogs :) No need to download anything on demand, no need to include any URL's in the emails :) All that requires extra code, this above really only requires one additional value
> 
> ...
> 
> > With the above change, I could see why it failed and make my own choice/decision before even running MerlinAU. I also wouldn't need to go above and beyond to manually download the changelog and review it manually. :)
> 
> ...
> 
> > Also it would of already sorted and only kept the propriate changes between the current and upcoming firmware, instead of all of the changes since the dawn of time haha
> 
> You seemed to have conflated the two separate features/events.
> 
> The new "`cl) View the latest F/W Changelog file.`" menu option implementation (as I suggested above) would be available all the time, regardless of any “high-risk phrases” and **independently** of any existing email notification setup that informs users of such “high-risk phrases” _**after**_ trying to flash the FW image. The menu option would allow users to review the Changelog at any time on-demand, before or after deciding to flash F/W image, and whether email notifications are set up by the user.
> 
> The 2nd feature/event is the email notification itself, which happens only **after** trying to flash the F/W image & only if email notifications are set up. During this event, the above menu option has nothing to do with the email. They are 2 completely separate & independent features, related only because they deal with the Changelog. Therefore, the notification email doesn’t need to download (again) the Changelog on-demand, so it would be simpler to provide the Changelog URL in the email itself so users can just click on it to review the latest changes listed at the top of the page. Yes, it includes previous versions, but that’s not a “bad” thing as it gives you more context of what came before & can compare the currently installed version vs the next update version (before deciding to continue to flash the new FW image).
> 
> Do you see what I mean by the 2 features/events being separate & independent of each other? In fact, you can modify the email notification contents now as you see fit without having the new menu option implemented.
> 
> We can talk more later tonight but I hope things are much clearer now.

No confusion or conflation at all on my part.
I always knew they were 2 different features lol.

As I said above here:
> I'm having a coffee and thinking, there is a limitation to doing it that way though. Even though it's the simplest, it does require you to have email notifications setup in AMTM and MerlinAU.
> 
> So naturally there would be that dependency else you'd still have to manually download to review the changelogs if high risk phrases are found.

And here:
> 
> The more I think about it, the more I like this solution even though it isn't simpler, why? because it's available all the time, even if the user hasn't setup email notifications. That removes the dependency to force people to set that up to review the changelogs.

I'm spit balling different design ideas.
Just because it's a suggestion doesn't mean it needs to be implemented that way.

My request was to see the changelogs somehow but I don't really care how. I hope that makes more sense now :)

**Comment by ExtremeFiretop** [comment]
Not sure where either gave the impression they are the same feature.

They do both however cover my concern 😂 so whichever floats our boats, as I also said here:



> I'm thinking I'll let you start how you see fit for the Logs Menu. We can always modify and adjust later, I'll keep sleeping on the changelog idea.



> I'm thinking I'll let you start how you see fit for the Logs Menu. We can always modify and adjust later, I'll keep sleeping on the changelog idea.

**Comment by ExtremeFiretop** [comment]
The point I was trying to make, I guess maybe I didn't make that clear, was that if we implement the changes in the email, there no no more need in my eyes to add an option on the menu available all the time.

I came to the conclusion the only reason to have it on the menu all the time would be due to the fact it's better sign it didn't require the email dependence to be setup. :)

**Comment by ExtremeFiretop** [comment]
> Really that's all I was suggesting for changelogs :) No need to download anything on demand, no need to include any URL's in the emails :) All that requires extra code, this above really only requires one additional value

With this line above, all I was trying to say is that suggestion I made was simpler than the 2 pervious ones made and still addressed my request to see the changelogs.

**Comment by Martinski4GitHub** [comment]
> Any ideas or thoughts on this one: https://www.snbforums.com/threads/merlinau-v1-1-2-the-ultimate-firmware-auto-updater-now-available-in-amtm.88577/post-904825
>
I've posted a response to that post, and submitted PR #203 as an attempt to address the issue.

**Comment by Martinski4GitHub** [comment]
> I'm thinking I'll let you start how you see fit for the Logs Menu. We can always modify and adjust later, I'll keep sleeping on the changelog idea.
>
OK, I'll start implementing a new menu option to view the F/W update logs. 

BTW, for my day job I've been tasked with writing a post-mortem report of our s/w development processes with a "lessons learned" analysis WRT to our latest production release. This means that I will be very busy for the next 3 working days since the report is due on Tuesday. I may have some time in the evenings to implement the menu option but it will likely be slow progress, but it should be OK since it's not something urgent. Just FYI.

**Comment by ExtremeFiretop** [comment]
> > I'm thinking I'll let you start how you see fit for the Logs Menu. We can always modify and adjust later, I'll keep sleeping on the changelog idea.
> 
> OK, I'll start implementing a new menu option to view the F/W update logs.
> 
> BTW, for my day job I've been tasked with writing a post-mortem report of our s/w development processes with a "lessons learned" analysis WRT to our latest production release. This means that I will be very busy for the next 3 working days since the report is due on Tuesday. I may have some time in the evenings to implement the menu option but it will likely be slow progress, but it should be OK since it's not something urgent. Just FYI.

No worries. Take your time, I was just thinking the take away for me is that I'd be happy with the bare minimum.
My original suggestion of creating a "Logs Menu" was only since were going to be playing with the log settings and we were gonna end up with a lot of them including the changelog I wanted to look into and the MerlinAU logs you were looking into.

But I don't need  a "logs menu" or really want one.. if we can avoid it :)

**Comment by Martinski4GitHub** [comment]
> 
> No worries. Take your time, I was just thinking the take away for me is that I'd be happy with the bare minimum. My original suggestion of creating a "Logs Menu" was only since were going to be playing with the log settings and we were gonna end up with a lot of them including the changelog I wanted to look into and the MerlinAU logs you were looking into.
> 
> But I don't need a "logs menu" or really want one.. if we can avoid it :) Like I said, I'll let you run with the design you think works best to start! :)
>
Yes, my thinking is to start small & simple with no "Logs Menu" for now. Later on, we can revisit the other ideas WRT the changelog. These are just "nice to have" features IMO, so there's no pressure to make a final decision at this point.

**Comment by ExtremeFiretop** [comment]
> > No worries. Take your time, I was just thinking the take away for me is that I'd be happy with the bare minimum. My original suggestion of creating a "Logs Menu" was only since were going to be playing with the log settings and we were gonna end up with a lot of them including the changelog I wanted to look into and the MerlinAU logs you were looking into.
> > But I don't need a "logs menu" or really want one.. if we can avoid it :) Like I said, I'll let you run with the design you think works best to start! :)
> 
> Yes, my thinking is to start small & simple with no "Logs Menu" for now. Later on, we can revisit the other ideas WRT the changelog. These are just "nice to have" features IMO, so there's no pressure to make a final decision at this point.

Fully agreed.

 I just didn't want you to think I was pushing any specific way. I was always just spit balling and throwing different ideas out to see what sticks.

Honestly any of the 2 suggestions I made, or any of the 2 you made, would deal with my request. So any of the 4 design choices would work Lool!

The thing that annoys me today is when high risk phrases are found. It stops the update over night sends me an email. And if I wake up and see this email, right now I gotta go to that URL or download the zip and review myself on the side outside of this script.

Ideally if it was in the email that reported it in the first place, or handled with a menu option all without leaving MerlinAU, then we are cooking with fire!

**Comment by Martinski4GitHub** [comment]
> Not sure where I gave the impression they are the same feature.
> 
> ...
> 
> I came back and said if we are discussion simplicity, my original "option 2" is the simplest as it didn't require any of the additional logic your option 1 or 2 required. Your first requires logic to download stuff on demand. And your second suggestion needed additional logic to detect which URL to send in the email.
> 
> In comparison my suggestion option 2, simply needed an additional variable and it handled my request. Not sure how this gave the impression I was confused or conflating.
> 
> And to be clear, there's nothing wrong with implementing both!
> 
> But my take away is if we were going for simplicity really my request is handled with just that simple change to the email template and Id be happy since I could see the changelog in the morning when I saw that notification without any additional steps required.
>

It seems to me that we had a major case of miscommunication here because something got "lost in translation."

When I said you were conflating the two features/events, I didn’t say you were confused. Those are not the same thing. The point was that, in my view, the suggested features were two separate ideas, but in the following paragraphs you appear to combine them as if they were one suggestion:

>
> Really that's all I was suggesting for changelogs :) No need to download anything on demand, no need to include any URL's in the emails :) All that requires extra code, this above really only requires one additional value
>
...
>
> With the above change, I could see why it failed and make my own choice/decision before even running MerlinAU. I also wouldn't need to go above and beyond to manually download the changelog and review it manually. :)
>

Now, after your later explanations, it became clear that’s not what you meant to convey, but that’s how the original messages came across without any additional context. So that’s why I thought I needed to explain and clarify that the two suggestions were to be taken separately & independently so we could discuss the benefits or disadvantages of each approach on their own.

Honestly, I got the impression that you were being defensive about my response clarifying the situation. I don’t really know why or what the trigger was since I assume you already understood how we communicate, and how often I need to have a clear & unambiguous understanding of things, especially before design/coding decisions. I could be wrong, but your writing style changes when you get defensive (based on past experience) and you might not be aware of it.

In any case, I really don't want to have another debate about this. I think we need to take a couple of steps back and reevaluate the options. Then, either we can drop this discussion if it’s no longer productive & we're not moving forward, or if you want later on we can revisit the suggested ideas (or present newer ideas) with much clearer heads and without the defensive postures. I have no problem either way. Miscommunications happen, but then we can clear the air and move forward.

And to be clear, I really have no preference WRT which of the solutions to view the Changelog gets implemented. I would be OK with whatever you decide.

**Comment by ExtremeFiretop** [comment]
> > Not sure where I gave the impression they are the same feature.
> > ...
> > I came back and said if we are discussion simplicity, my original "option 2" is the simplest as it didn't require any of the additional logic your option 1 or 2 required. Your first requires logic to download stuff on demand. And your second suggestion needed additional logic to detect which URL to send in the email.
> > In comparison my suggestion option 2, simply needed an additional variable and it handled my request. Not sure how this gave the impression I was confused or conflating.
> > And to be clear, there's nothing wrong with implementing both!
> > But my take away is if we were going for simplicity really my request is handled with just that simple change to the email template and Id be happy since I could see the changelog in the morning when I saw that notification without any additional steps required.
> 
> It seems to me that we had a major case of miscommunication here because something got "lost in translation."
> 
> When I said you were conflating the two features/events, I didn’t say you were confused. Those are not the same thing. The point was that, in my view, the suggested features were two separate ideas, but in the following paragraphs you appear to combine them as if they were one suggestion:
> 
> > Really that's all I was suggesting for changelogs :) No need to download anything on demand, no need to include any URL's in the emails :) All that requires extra code, this above really only requires one additional value
> 
> ...
> 
> > With the above change, I could see why it failed and make my own choice/decision before even running MerlinAU. I also wouldn't need to go above and beyond to manually download the changelog and review it manually. :)
> 
> Now, after your later explanations, it became clear that’s not what you meant to convey, but that’s how the original messages came across without any additional context. So that’s why I thought I needed to explain and clarify that the two suggestions were to be taken separately & independently so we could discuss the benefits or disadvantages of each approach on their own.
> 
> Honestly, I got the impression that you were being defensive about my response clarifying the situation. I don’t really know why or what the trigger was since I assume you already understood how we communicate, and how often I need to have a clear & unambiguous understanding of things, especially before design/coding decisions. I could be wrong, but your writing style changes when you get defensive (based on past experience) and you might not be aware of it.
> 
> In any case, I really don't want to have another debate about this. I think we need to take a couple of steps back and reevaluate the options. Then, either we can drop this discussion if it’s no longer productive & we're not moving forward, or if you want later on we can revisit the suggested ideas (or present newer ideas) with much clearer heads and without the defensive postures. I have no problem either way. Miscommunications happen, but then we can clear the air and move forward.
> 
> And to be clear, I really have no preference WRT which of the solutions to view the Changelog gets implemented. I would be OK with whatever you decide.

Agreed. 

Thanks for clearing the air. It's much appreciated. 👍 :) I promise wasn't trying to be defensive. I'm happy I was able to clarify that. 

Sometimes I think the communication would be best if we had a call once a week or something, but I realize that may be outside of our control with the time difference!

I'm happy with any choice, my goal was to not make this a burden because I'd be happy with literally the base minimum implementation hahah 🤣

**Comment by Martinski4GitHub** [comment]
> > > No worries. Take your time, I was just thinking the take away for me is that I'd be happy with the bare minimum. My original suggestion of creating a "Logs Menu" was only since were going to be playing with the log settings and we were gonna end up with a lot of them including the changelog I wanted to look into and the MerlinAU logs you were looking into.
> > > But I don't need a "logs menu" or really want one.. if we can avoid it :) Like I said, I'll let you run with the design you think works best to start! :)
> > 
> > 
> > Yes, my thinking is to start small & simple with no "Logs Menu" for now. Later on, we can revisit the other ideas WRT the changelog. These are just "nice to have" features IMO, so there's no pressure to make a final decision at this point.
> 
> Fully agreed.
> 
> I just didn't want you to think I was pushing any specific way. I was always just spit balling and throwing different ideas out to see what sticks.
> 
> Honestly any of the 2 suggestions I made, or any of the 2 you made, would deal with my request. So any of the 4 design choices would work Lool!
> 
> The thing that annoys me today is when high risk phrases are found. It stops the update over night sends me an email. And if I wake up and see this email, right now I gotta go to that URL or download the zip and review myself on the side outside of this script.
> 
> Ideally if it was in the email that reported it in the first place, or handled with a menu option all without leaving MerlinAU, then we are cooking with fire!
>
I'll let you take the lead on this feature and decide what works best from a user's perspective. My own perspective is completely different because I have a more hands-on, technical-oriented approach to how I review the Changelog & monitor user feedback before I decide when to flash the latest F/W image. But as mentioned, I'm OK with whatever you decide.

**Comment by Martinski4GitHub** [comment]
> 
> Agreed.
> 
> Thanks for clearing the air. It's much appreciated. 👍 :) I promise wasn't trying to be defensive. I'm happy I was able to clarify that.
> 
> Sometimes I think the communication would be best if we had a call once a week or something, but I realize that may be outside of our control with the time difference!
> 
Several studies have pretty much established that communication works best when in person face-to-face, not only because of the interpersonal relationships that grow organically but also because we can see the other person's face and pick up cues from their body language, and their voice tone & modulation also conveys meaning. All this gives you the full "measure" of what the person is trying to say (assuming, of course, that no one is trying to deceive the other). The 2nd best method is via teleconference (i.e. voice or Zoom meetings).

Anyway, I understand what you mean but a weekly call is not something I'd be willing to do even if we were in the same timezone. I already have enough meetings as it is (in person & teleconferences) for my day job, and I'm not fond of meetings in the first place, but they are a necessary "burden." Also, this project is just a hobby for me, and having weekly calls would feel like another extra job & a burden to schedule in my off-hours. Nothing personal, of course - it's just my personal view :>).

So we'll just keep communicating as usual via written exchanges while keeping in mind that sometimes what we write does not always convey what we are trying to say. Clear & effective written communication is an art form all by itself, and it's extremely difficult to fully master it. I still have much more to learn in that regard.

**Comment by ExtremeFiretop** [comment]
> I'll let you take the lead on this feature and decide what works best from a user's perspective. My own perspective is completely different because I have a more hands-on, technical-oriented approach to how I review the Changelog & monitor user feedback before I decide when to flash the latest F/W image. But as mentioned, I'm OK with whatever you decide.

Hey @Martinski4GitHub 
Created PR: https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/204

**This includes the suggestion you made below:**

> If I understand the desired end-result, wouldn't it be simpler to just provide a menu option to download the appropriate Changelog "on demand" **without** actually downloading any F/W ZIP file?
> 
> For example, we add a new menu option like: `cl) View the latest F/W Changelog file.`
> 
> When the user enters "**cl**" the script would download the appropriate data (via **curl** cmd) using the corresponding URL.

**As the agreed upon method:**

> The more I think about it, the more I like this solution even though it isn't simpler, why? because it's available all the time, even if the user hasn't setup email notifications. That removes the dependency to force people to set that up to review the changelogs.
